torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -125,9 +132,77 @@ class ProcessGroupNCCL : public ProcessGroup {
     // to the store.
     std::shared_ptr<Store> store_;
 
+    // Store a Future work associated with WorkNCCL.
+    c10::intrusive_ptr<c10::ivalue::Future> futureWork_;
+
     friend class ProcessGroupNCCL;
   };
 
+  // FutureNCCL is a subclass of ivalue's Future. The goal is to use
+  // this class in getFuture API of WorkNCCL. This Future is mostly a
+  // wrapper to synchronize streams appropriately and it mostly enables
+  // the async programming model of CUDA while trying to adhere to the
+  // Future interface.
+  //
+  // FutureNCCL has a reference to WorkNCCL and NCCL collective's outputs.
+  // Its value is NCCL collective's outputs.
+  struct FutureNCCL : at::ivalue::Future {
+   public:
+    explicit FutureNCCL(
+        std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work,
+        std::vector<at::Tensor>& outputs)
+        : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
+          work_(work),
+          outputs_(outputs) {}
+
+    // Simply calls WorkNCCL's wait(). It will return after synchronizing
+    // the correct GPU streams to ensure we can have async CUDA execution
+    // and it does not wait for the entire operation to complete on GPU.
+    // If NCCL_BLOCKING_WAIT is enabled, in that case, it will wait for the
+    // entire operation to complete before returning.
+    void wait() override {
+      work_->wait();
+    }
+
+    // FutureNCCL's value is NCCL collective's outputs and callbacks were
+    // invoked inline by addCallback(), so markCompleted is not needed.
+    void markCompleted(at::IValue /* unused */) override {
+      C10_THROW_ERROR(Error, "FutureNCCL::markCompleted is not supported.");
+    }
+
+    // Returns NCCL collective's outputs after WorkNCCL wait() returns.
+    at::IValue value() override {
+      work_->wait();
+      return outputs_;
+    }
+
+    const at::IValue& constValue() override {
+      work_->wait();
+      return outputs_;
+    }
+
+    // Add a callback to FutureNCCL. FutureNCCL invokes the callback inline.
+    // Callbacks return a Future (not FutureNCCL).
+    void addCallback(std::function<void(void)> callback) override {
+      callback();
+    }
+
+    // Checks whether NCCL work is completed.
+    bool completed() const override {
+      return work_->isCompleted();
+    }
+
+    // FutureNCCL has a value that was set in its constructor as NCCL
+    // collective's outputs.
+    bool hasValue() const override {
```

Comment:
@pritamdamania87 @mrshenli I followed @wanchaol's comment and return `completed` inside `hasValue`. The only problem is outputs could be ready and accessed by reducer or callback before `completed` is true since it checks for `work->isCompleted()`. I think that's fine though.

Reply:
> The only problem is outputs could be ready and accessed by reducer or callback before completed is true since it checks for work->isCompleted(). 

Could you please elaborate? Why the outputs can be ready before the NCCL work is completed?

Reply:
@wanchaol @mrshenli FutureNCCL is a wrapper around c10d work to ensure we can adhere to the Future's API and at the same time support the async model that CUDA provides. In the current version, you can call `value()` before the work has actually completed on the GPU and use that value in another operation. This is possible due to the async nature of CUDA and the fact that we synchronize the streams correctly. Since, you can always call `value()` irrespective of whether or not the work has completed on the GPU, I think for consistency `hasValue()` should always return True.

Reply:
> > The only problem is outputs could be ready and accessed by reducer or callback before completed is true since it checks for work->isCompleted().
> 
> Could you please elaborate? Why the outputs can be ready before the NCCL work is completed?

@mrshenli By default `NCCL_BLOCKING_WAIT=0`, so `work->wait()` will just return after synchronizing streams, whereas `work->isCompleted()` still checks for cuda success.

Reply:
@mrshenli @wanchaol reverted the change on `hasValue` after @pritamdamania87's comment. I think it makes sense to simply return true, since we set the value right at the constructor and always wait inside `FutureNCCL::value`.